File: ios/LdkFilter.swift
//
//  LdkFilter.swift
//  react-native-ldk
//
//  Created by Jason van den Berg on 2022/05/10.
//

import Foundation
import LDKFramework

class LdkFilter: Filter {
    override func register_tx(txid: [UInt8]?, script_pubkey: [UInt8]) {
        let txIdHex = Data(txid ?? []).hexEncodedString()
        let scriptPubkeyHex = Data(script_pubkey).hexEncodedString()
    
        sendEvent(eventName: .register_tx, eventBody: ["txid": txIdHex, "script_pubkey": scriptPubkeyHex])
    }
    
    override func register_output(output: WatchedOutput) -> Option_C2Tuple_usizeTransactionZZ {
        let scriptPubkeyHex = Data(output.get_script_pubkey()).hexEncodedString()
        let outpoint = output.get_outpoint()!
        let blockHashBytes = Data(output.get_block_hash()).hexEncodedString()
        sendEvent(eventName: .register_output, eventBody: ["block_hash": blockHashBytes, "index": String(outpoint.get_index()), "script_pubkey": scriptPubkeyHex])
        
        return Option_C2Tuple_usizeTransactionZZ.none()
    }
}

File: lib/android/src/main/java/com/reactnativeldk/classes/LdkFilter.kt
package com.reactnativeldk.classes

import com.facebook.react.bridge.Arguments
import com.reactnativeldk.EventTypes
import com.reactnativeldk.LdkEventEmitter
import com.reactnativeldk.hexEncodedString
import com.reactnativeldk.putHexString
import org.ldk.structs.Filter
import org.ldk.structs.Filter.FilterInterface
import org.ldk.structs.Option_ThirtyTwoBytesZ
import org.ldk.structs.WatchedOutput

class LdkFilter {
    var filter = Filter.new_impl(object : FilterInterface {
        override fun register_tx(txid: ByteArray, script_pubkey: ByteArray) {
            val body = Arguments.createMap()
            body.putString("txid", txid.reversedArray().hexEncodedString())
            body.putString("script_pubkey", script_pubkey.hexEncodedString())
            LdkEventEmitter.send(EventTypes.register_tx, body)
        }

        override fun register_output(output: WatchedOutput) {
            val body = Arguments.createMap()
            if (output._block_hash is Option_ThirtyTwoBytesZ.Some) {
                body.putHexString("block_hash", (output._block_hash as Option_ThirtyTwoBytesZ.Some).some)
            }
            body.putInt("index", output._outpoint._index.toInt())
            body.putString("script_pubkey", output._script_pubkey.hexEncodedString())
            LdkEventEmitter.send(EventTypes.register_output, body)
        }
    })
}